zookeeper/zookeeper.go

```diff
@@ -0,0 +1,204 @@
+		parser:             opts.ConfigParser,
+		serverPathTemplate: serverPathTemplate,
+		clientPathTemplate: clientPathTemplate,
+		prefixTemplate:     prefixTemplate,
+	}
+	return c, nil
+}
+
+// SetParser support customise parser
+func (c *client) SetParser(parser ConfigParser) {
+	c.parser = parser
+}
+
+func (c *client) render(cpc *ConfigParamConfig, t *template.Template) (string, error) {
+	var tpl bytes.Buffer
+	err := t.Execute(&tpl, cpc)
+	if err != nil {
+		return "", err
+	}
+	return tpl.String(), nil
+}
+
+// ServerConfigParam render server config parameters
+func (c *client) ServerConfigParam(cpc *ConfigParamConfig) (ConfigParam, error) {
+	return c.configParam(cpc, c.serverPathTemplate)
+}
+
+// ClientConfigParam render client config parameters
+func (c *client) ClientConfigParam(cpc *ConfigParamConfig) (ConfigParam, error) {
+	return c.configParam(cpc, c.clientPathTemplate)
+}
+
+// configParam render config parameters. All the parameters can be customized with CustomFunction.
+// ConfigParam explain:
+//  1. Prefix: /KitexConfig by default.
+//  2. ServerPath: {{.ServerServiceName}}.{{.Category}} by default.
+//     ClientPath: {{.ClientServiceName}}.{{.ServerServiceName}}.{{.Category}} by default.
+func (c *client) configParam(cpc *ConfigParamConfig, t *template.Template) (ConfigParam, error) {
+	param := ConfigParam{}
+	var err error
+	param.Path, err = c.render(cpc, t)
+	if err != nil {
+		return param, err
+	}
+	param.Prefix, err = c.render(cpc, c.prefixTemplate)
+	if err != nil {
+		return param, err
+	}
+	return param, nil
+}
+
+// DeregisterConfig deregister the config.
+func (c *client) DeregisterConfig(path string, uniqueID int64) {
+	mu.Lock()
+	clientKey := path + "/" + strconv.FormatInt(uniqueID, 10)
+	cancel := ctxMap[clientKey]
+	cancel()
+	mu.Unlock()
+}
+
+// RegisterConfigCallback register the callback function to zookeeper client.
+func (c *client) RegisterConfigCallback(ctx context.Context, path string, uniqueID int64, callback func(string, ConfigParser)) {
+	clientCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		mu.Lock()
+		clientKey := path + "/" + strconv.FormatInt(uniqueID, 10)
+		ctxMap[clientKey] = cancel
+		mu.Unlock()
+		for {
+			_, _, watchChan, err := c.zConn.ExistsW(path)
+			if err != nil {
+				klog.Debugf("[zookeeper] watch node %s failed %v", path, err)
+				return
+			}
+			select {
+			case <-clientCtx.Done():
+				return
+			case watchEvent := <-watchChan:
+				if watchEvent.Type == zk.EventNodeDataChanged {
+					data, _, err := c.zConn.Get(path)
+					nodeValue := string(data)
+					klog.Debugf("[zookeeper] config path: %s updated,data is %s", path, nodeValue)
+					if err != nil {
+						klog.Warnf("[zookeeper] get config %s from zookeeper failed: %v", path, err)
+					}
+					callback(nodeValue, c.parser)
+				} else if watchEvent.Type == zk.EventNodeDeleted {
+					klog.Debugf("[zookeeper] config path: %s deleted", path)
+					callback("", c.parser)
```

Comment:
It's possible for user to set the value to be "", so it's better we distinguish it. Please refer to https://github.com/kitex-contrib/config-etcd/pull/2/files for more details.

Reply:
Done